### PR TITLE
cve-update-nvd2-native: Add database file pre-downloading feature

### PR DESCRIPTION
**Pull request by masami256**
# Purpose of pull request

This commit adds the feature to download a prepared database file to reduce
access to the NVD's API server. Typically, cve-update-nvd2-native creates a
database file using the NVD JSON API. However, it takes time to retrieve all
CVE information via the API. Therefore, it would be beneficial to reduce this
time. Additionally, the NVD API server can sometimes be unstable for short or
long periods. In such cases, downloading a prepared database file instead of
creating it from scratch would be preferable.

To download the file from the server, we must check its checksum as per BitBake's
policy. However, since the file may be updated, we cannot determine its
checksum in advance. Therefore, we temporarily disable the checksum check during
the download. Even with the checksum check disabled, BitBake will show a warning
message, which can be ignored.

```
WARNING: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Missing md5 SRC_URI checksum for /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db, consider adding to the recipe:
SRC_URI[md5sum] = "30e3069969f28e6cf07ceb6645267d56"
WARNING: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Missing sha256 SRC_URI checksum for /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db, consider adding to the recipe:
SRC_URI[sha256sum] = "e9540629d955a0f045c0becd5f02b271dd28062dc0b649c9c30840187562c8e2"
```

To download the file, the user needs the following setup in local.conf or
another config file. The download file name must be nvdcve_2.db.

```
CVE_DB_PREDOWNLOAD="1"
CVE_DB_PREDOWNLOAD_URL="http://somewhere/nvdcve_2.db"
```

We pre-download the database file when CVE_DB_PREDOWNLOAD is set to 1 and one of the following conditions is met:
 - The database file is not present
 - The database file is present, and the modification time (mtime) is too old

# Test
### Prepare

1. Run cve-check if you don't  have nvdcve_2.db
2. copy $DL_DIR/CVE_CHECK/nvdcve_2.db to other directory(e.g. $HOME/tmp)
3. Run "cd $HOME/tmp"
4. Run " python3 -m http.server  8080"

### Test1

No nvdcve_2.db file and download prepared database file.
Then update database.

1. Remove $DL_DIR/CVE_CHECK/nvdcve_2.db
2. Add following lines in your local.conf (Also add NVDCVE_API_KEY value if you have NVD api key)

```
CVE_DB_PREDOWNLOAD="1"
CVE_DB_PREDOWNLOAD_URL="http://localhost:8080/nvdcve_2.db"
```

2. Run "bitbake cve-update-nvd2-native -vv"

### Test2

nvdcve_2.db exist and download prepared database file then update database.

1. Ensure $DL_DIR/CVE_CHECK/nvdcve_2.db is present
2. Add 'CVE_DB_UPDATE_INTERVAL = "10"' to local.conf
3. Run "bitbake cve-update-nvd2-native -vv"

### Test3

Create database using NVD api (default behavior when run cve-check first time)

1. Remove $DL_DIR/CVE_CHECK/nvdcve_2.db
2. Remove CVE_DB_PREDOWNLOAD, CVE_DB_PREDOWNLOAD_URL, and CVE_DB_UPDATE_INTERVAL from local.conf
3. Run "bitbake cve-update-nvd2-native -vv"

### Test1

It download database file from web server, then update CVE entries.

```
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: There is no NVD database file, we will download it.
Fetching prepared NVD database file from ... http://localhost:8080/nvdcve_2.db;downloadfilename=downlaod_temp_nvdcve_2.db
WARNING: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Missing md5 SRC_URI checksum for /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db, consider adding to the recipe:
SRC_URI[md5sum] = "30e3069969f28e6cf07ceb6645267d56"
WARNING: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Missing sha256 SRC_URI checksum for /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db, consider adding to the recipe:
SRC_URI[sha256sum] = "e9540629d955a0f045c0becd5f02b271dd28062dc0b649c9c30840187562c8e2"
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: file renamed to /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db to /home/build/work/emlinux/latest-dev/build/../downloads/CVE_CHECK/nvdcve_2.db
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: CVE database: performing partial update
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Updating entries
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Requesting https://services.nvd.nist.gov/rest/json/cves/2.0?startIndex=0&lastModStartDate=2024-07-05T05%3A38%3A20%2B00%3A00&lastModEndDate=2024-07-08T04%3A44%3A23.486207%2B00%3A00
Currently  1 running tasks (1 of 4)   0% |       
```

### Test2

It download database file even though database file exists. Then update CVE entries.

```
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Download and replace NVD database file. 
Fetching prepared NVD database file from ... http://localhost:8080/nvdcve_2.db;downloadfilename=downlaod_temp_nvdcve_2.db
WARNING: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Missing md5 SRC_URI checksum for /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db, consider adding to the r
ecipe:                                                                                             
SRC_URI[md5sum] = "30e3069969f28e6cf07ceb6645267d56"                                            
WARNING: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Missing sha256 SRC_URI checksum for /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db, consider adding to th
e recipe:                                                                                          
SRC_URI[sha256sum] = "e9540629d955a0f045c0becd5f02b271dd28062dc0b649c9c30840187562c8e2"         
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: file renamed to /home/build/work/emlinux/latest-dev/build/../downloads/downlaod_temp_nvdcve_2.db to /home/build/work/emlinux/latest-dev/build/
../downloads/CVE_CHECK/nvdcve_2.db                                                                 
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: CVE database: performing partial update 
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Updating entries                        
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Requesting https://services.nvd.nist.gov/rest/json/cves/2.0?startIndex=0&lastModStartDate=2024-07-05T05%3A38%3A20%2B00%3A00&lastModEndDate=202
4-07-08T04%3A20%3A44.296292%2B00%3A00                                                              
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Got 275 entries   
```

### Test3

It executes NVD API to get CVE information. As you can see "startIndex=0" which means get oldest data

Create
```
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: CVE database: performing partial update 
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Updating entries                        
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Requesting https://services.nvd.nist.gov/rest/json/cves/2.0?startIndex=0&lastModStartDate=2024-07-05T05%3A38%3A20%2B00%3A00&lastModEndDate=202
4-07-08T04%3A20%3A44.296292%2B00%3A00                                                              
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Got 275 entries   

default
NOTE: Executing SetScene Tasks
NOTE: Executing RunQueue Tasks
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Create CVE database using NVD API.
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Updating entries
NOTE: cve-update-nvd2-native-1.0-r0 do_populate_cve_db: Requesting https://services.nvd.nist.gov/rest/json/cves/2.0?startIndex=0
```

### CODE REVIEW

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 1]
```diff
@@ -37,18 +37,55 @@ CVE_DB_TEMP_FILE ?= "${CVE_CHECK_DB_DIR}/temp_nvdcve_2.db"
 
 CVE_CHECK_DB_FILE ?= "${CVE_CHECK_DB_DIR}/nvdcve_2.db"
 
+CVE_DB_DOWNLOAD_TEMP_FILE = "downlaod_temp_nvdcve_2.db"
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: This should be "CVE_DB_PREDOWNLOAD_TEMP_FILE".

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 2]
```diff
@@ -37,18 +37,55 @@ CVE_DB_TEMP_FILE ?= "${CVE_CHECK_DB_DIR}/temp_nvdcve_2.db"
...
+
 python () {
     if not bb.data.inherits_class("cve-check", d):
         raise bb.parse.SkipRecipe("Skip recipe when cve-check class is not loaded.")
 }
 
+def run_download_prepared_db_file(d):
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: This should be "run_predownload_db_file".

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 3]
```diff
@@ -37,18 +37,55 @@ CVE_DB_TEMP_FILE ?= "${CVE_CHECK_DB_DIR}/temp_nvdcve_2.db"
...
+    base_url = d.getVar("NVDCVE_PREPARED_DATABASE_URL")
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: This should be "CVE_DB_PREDOWNLOAD_URL" because this is not related with the services provided by NVD.

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 4]
```diff
@@ -37,18 +37,55 @@ CVE_DB_TEMP_FILE ?= "${CVE_CHECK_DB_DIR}/temp_nvdcve_2.db"
...
+    if len(base_url) == 0:
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: d.getVar() may return None, then the len(base_url) causes TypeError. We can use "if not base_url" instead.

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 5]
```diff
@@ -59,7 +96,6 @@ python do_populate_cve_db() {
     cleanup_db_download(db_file, db_tmp_file)
     # By default let's update the whole database (since time 0)
     database_time = 0
-
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: Is this removal unnecessary?

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 6]
```diff
@@ -68,11 +104,33 @@ python do_populate_cve_db() {
...
+
+            if need_download_db_file:
+                bb.note("Download and replace NVD database file.")
+                # We need update database file because database file is old.
+                if not download_prepared_db_file(d):
+                    return
+
+            database_time = os.path.getmtime(db_file)
+        else:
+            bb.note("Create CVE database using NVD API.")
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb:125]: I think we can simplify the above code.

```
     try:
         import time
         update_interval = int(d.getVar("CVE_DB_UPDATE_INTERVAL"))
         if update_interval < 0:
             bb.note("CVE database update skipped")
             return
+        elif d.getVar("CVE_DB_PREDOWNLOAD") == "1":
+            if not os.path.exists(db_file):
+                bb.note("There is no CVE database file locally, we will pre-download it.")
+                if not download_prepared_db_file(d):
+                    return
+            elif time.time() - os.path.getmtime(db_file) >= update_interval:
+                bb.note("Too old CVE database file found locally, we will replace it by pre-download.")
+                if not download_prepared_db_file(d):
+                    return
         if time.time() - os.path.getmtime(db_file) < update_interval:
             bb.note("CVE database recently updated, skipping")
             return
     except OSError:
         pass
```

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 7]
```diff
@@ -68,11 +104,33 @@ python do_populate_cve_db() {
         if update_interval < 0:
             bb.note("CVE database update skipped")
             return
-        if time.time() - os.path.getmtime(db_file) < update_interval:
-            bb.note("CVE database recently updated, skipping")
-            return
-        database_time = os.path.getmtime(db_file)
 
+        need_download_db_file = d.getVar("NVDCVE_DOWNLOAD_PREPARED_DATABASE") == "1"
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: This should be "CVE_DB_PREDOWNLOAD" instead of "NVDCVE_DOWNLOAD_PREPARED_DATABASE". (again, this is not related with the services provided by NVD.)

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 8]
```diff
@@ -37,18 +37,55 @@ CVE_DB_TEMP_FILE ?= "${CVE_CHECK_DB_DIR}/temp_nvdcve_2.db"
...
+        bb.plain('Fetching prepared NVD database file from ... %s' % url)
+        fetcher = bb.fetch2.Fetch([url], d)
+        fetcher.download()
+    except bb.fetch2.BBFetchException as e:
+        bb.warn("Failed to fetch prepared NVD database file: %s" % e)
+        return False
+
+    return True
+
+def download_prepared_db_file (d):
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb]: This should be "predownload_db_file".

**Comment by hiraku-wfs** [comment]
Thank you for your proposal. I think your approach is good.

Please use "database file pre-downloading" for the feature name, instead of "download prepared database file".
So commit (and PR) title should be "cve-update-nvd2-native: Add database file pre-downloading feature".

I will follow this comment with review comments.

In addition, could you please polish your commit message? It looks like there are a lot of typos and other small errors.

File: recipes-core/cve-update/cve-update-nvd2-native.bb (hunk 8 above)
Comment (masami256) [reply recipes-core/cve-update/cve-update-nvd2-native.bb]: @hiraku-wfs all fixed and pushed new commit.

File: recipes-core/cve-update/cve-update-nvd2-native.bb [hunk 9]
```diff
@@ -68,11 +105,25 @@ python do_populate_cve_db() {
         if update_interval < 0:
             bb.note("CVE database update skipped")
             return
-        if time.time() - os.path.getmtime(db_file) < update_interval:
-            bb.note("CVE database recently updated, skipping")
-            return
-        database_time = os.path.getmtime(db_file)
 
+        elif d.getVar("CVE_DB_PREDOWNLOAD") == "1":
```
Comment (hiraku-wfs) [inline recipes-core/cve-update/cve-update-nvd2-native.bb:108]: Could you remove the above empty line (Line 108) or change the `elif` to `if`?

I think this is the last item I point out for the code.

**Comment by hiraku-wfs** [comment]
Please modify the commit message as follows (and also update this PR description).

```
- To download a file from the server, 
+ To download the file from a server, 
```

```
- CVE_DB_PREDOWNLOAD=="1"
- CVE_DB_PREDOWNLOAD_URL=="http://somewhere/nvdcve_2.db"
+ CVE_DB_PREDOWNLOAD="1"
+ CVE_DB_PREDOWNLOAD_URL="http://somewhere/nvdcve_2.db"
```

```
- We download the database file when one of the following conditions is met:
- - The database file is not present, and CVE_DB_PREDOWNLOAD is set to 1
- - The database file is present, CVE_DB_PREDOWNLOAD_URL is set to 1, and the database file's modification time (mtime) is old
+ We pre-download the database file when CVE_DB_PREDOWNLOAD is set to 1 and one of the following conditions is met:
+ - The database file is not present
+ - The database file is present, and the modification time (mtime) is too old
```

File: recipes-core/cve-update/cve-update-nvd2-native.bb (hunk 9 above)
Comment (masami256) [reply recipes-core/cve-update/cve-update-nvd2-native.bb:108]: @hiraku-wfs  I pushed new commit.

**Comment by hiraku-wfs** [comment]
LGTM for the code.
Could you update "How to test" and "Test result"?

**Comment by hiraku-wfs** [comment]
Thanks!
